src/collectors/ceph/ceph.py

```diff
@@ -69,78 +169,357 @@ def get_default_config(self):
-                          name, err)
-            self.log.exception('Could not get stats from %s' % name)
-            return {}
+        # name of <metric>
+        def _make_stat_name(stat_name):
+            parts = [counter_prefix] + path + [stat_name]
+            clean_parts = filter(None, parts)
+            return name_class(_PATH_SEP.join(clean_parts))
 
-        try:
-            json_data = json.loads(json_blob)
-        except Exception, err:
-            self.log.info('Could not parse stats from %s: %s',
-                          name, err)
-            self.log.exception('Could not parse stats from %s' % name)
-            return {}
+        total_sum_name = _make_stat_name("sum")
+        total_count_name = _make_stat_name("count")
+        delta_sum_name = _make_stat_name("delta_sum")
+        delta_count_name = _make_stat_name("delta_count")
+        delta_avg_name = _make_stat_name("last_interval_avg")
+
+        # lookup raw metric component values
+        total_sum = lookup_dict_path(stats, path, ['sum'])
+        total_count = lookup_dict_path(stats, path, ['avgcount'])
+
+        # perform metric-specific type conversions
+        if stat_type & _PERFCOUNTER_TIME:
+            total_sum = self._ceph_time_to_seconds(total_sum)
+
+        # Calculate deltas since last time we queried admin socket. The
+        # derivitive function records from the last invocation the
+        # total_sum/total_count, and simply returns the difference.
+        delta_sum = self.derivative(delta_sum_name, total_sum,
+                                    time_delta=False)
+        delta_count = self.derivative(delta_count_name, total_count,
+                                      time_delta=False)
+
+        # average in the last collection interval
+        if delta_count == 0:
+            delta_avg = 0
+        else:
+            delta_avg = float(delta_sum) / float(delta_count)
+
+        # publish raw data
+        if self.config['long_running_detail']:
+            self.publish_gauge(total_sum_name, total_sum)
+            self.publish_gauge(total_count_name, total_count)
+
+        # publish averages
+        self.publish_gauge(delta_avg_name, delta_avg, 6)
 
-        return json_data
+    def _ceph_time_to_seconds(self, val):
+        """Convert Ceph time format into seconds.  Older Ceph
+           versions output times as a string, while newer
+           versions output a float (which we pass through)
 
-    def _publish_stats(self, counter_prefix, stats):
-        """Given a stats dictionary from _get_stats_from_socket,
-        publish the individual values.
+        :param val: string in format "seconds.nanoseconds" or
+                    floating point number.
+
+        Returns:
+            Time in seconds as a floating point number.
         """
+        if isinstance(val, basestring):
+            sec, nsec = map(lambda v: long(v), val.split("."))
+            return float(sec * _NSEC_PER_SEC + nsec) / float(_NSEC_PER_SEC)
```

Comment:
I don't think there's a change needed here.  The input value is in <seconds>.<nanoseconds> format.  So if you had "1.1", that would *not* be equal to 1.1 seconds.  It would be equal to 1.000000001 -- the existing code is correct.